### ISSUE TITLE
A repeated call of signal handler and cause a stack exhaustion 

### ISSUE THREAD

**Issue by wangjianchaokernel** [systemd/systemd#5267]
### Submission type

  - [ x ] Bug report
  - [ ] Request for enhancement (RFE)

*NOTE: Do not submit anything other than bug reports or RFEs via the issue tracker!*

### systemd version the issue has been seen with

> All

*NOTE: Do not submit bug reports about anything but the two most recently released systemd versions upstream!*

The issue just appear on our system one time and we didn't get symbols information in  the core file.
the bt in gdb is like this:
...
#8240 < signal handler called >
#8241 0x0000000000000000 in ?? ()
...
I found a risk in exec_child() that may cause similar issue, rename_process_from_path() is invoked before default_signals(). 
If there is anything wrong occur in rename_process_from_path() and trigger a signal defined in SIGNALS_CRASH_HANDLER, the crash() will be invoked repeatedly

**Comment by poettering** [comment]
Well, that's a pretty theoretic weakness...

**Comment by wangjianchaokernel** [comment]
@poettering yes, but that seems to happen in our system. In the backtrace,  we could see more than 8000 repeated frames which repeat < signal handler caller > .

**Comment by evverx** [comment]
> the crash() will be invoked repeatedly

This looks like https://github.com/systemd/systemd/pull/3314

Which version of systemd are you using?

**Comment by evverx** [comment]
Oh, sorry. I was wrong. This issue unrelated to #3314 of course

**Comment by wangjianchaokernel** [comment]
@poettering , there seems to be a weakness in crash():
    if (getpid() != 1)
                /* Pass this on immediately, if this is not PID 1 */
               raise(sig);
If the crash() is invoked in a process whose pid is not 1, it will raise the signal immediately.
However, with SA_NODEFER flag set,  the crash() will be invoked again, because the signal handler is still crash() here.
Thus, we should invoke default_signal() before raise() the signal gain, or other similar operation here.
